tt_metal/detail/tt_metal.hpp

```diff
@@ -442,93 +446,146 @@ namespace tt::tt_metal{
             );
         }
 
-        inline void SendDispatchKernelsToDevice(Device *device, const SystemMemoryManager& manager, const uint32_t hugepage_channel) {
+        inline void SendDispatchKernelsToDevice(Device *device) {
```

Comment:
running dispatch on remote device requires having kernels loaded onto the connected mmio device (these pull and push from the remote command queue)

`SendDispatchKernelsToDevice(device)` is called when `device` is being initialized. 

Since we don't have a handle on mmio device when initializing the remote device, mmio device initialization is responsible for loading the remote CQ interfacing kernels 

When we have a device pool and can access device by ID this can be refactored to load dispatch kernels to service the particular `device` being initialized. 

FYI @TT-billteng @DrJessop

Reply:
I will have some apis that are at cluster level similar to this, assumes that mmio device initialized.

Reply:
@TT-billteng while trying to move dispatch_core_manager to be owned by Device, I realized the original reason I made it globally accessible via `device_id` was because during dispatch program creation for the MMIO device (which is a step of the MMIO device initialization) I don't have access to the remote device but I need to get the locations of the remote device's issue and completion queue interface cores. These cores are on the MMIO device. 

When there is a device pool and I can get access to the device by device id I can uplift the dispatch_core_manager to be owned by device and do the refactor of the dispatch program I mentioned above that would result in remote device initialization generating two programs for remote dispatch 
- program 0 would target MMIO device and include the issue and completion queue kernels
- program 1 would target remote device and include the remaining dispatch kernels

Reply:
@DrJessop this is similar to what @aliuTT  and I were mentioning the other day with remote dispatch being dependent on mmio device initialization. With your changes we will also need the command queue for the mmio device to be created because that the logic to send launch msg to dispatch cores + write L1 pointers is part of command queue creation. 

Dependence on mmio CQ can be removed once we can create the 2 programs I listed above during remote device initialization.

Reply:
ok sounds good, I always try to design using the minimum # of global variables as possible, but seems like we need it here